docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/maintenance.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Node Maintenance
+
+## Abstract
+
+A proposal for implementing safe node maintenance primitives, in stages.
+
+## Overview
+
+Nodes need periodic maintenance, most of which require partial or total
+unavailability to pods.  The cluster admin would like for this to happen (1)
+automatically, and (2) respecting any service quality constraints.
+
+Examples of maintenance:
+
+- unmount/remount filesystems to change mkfs params
+- remove from service but leave up for debugging
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- remove from service permanently
+
+## Service Quality
+
+This may be thought of as the cell admin offering a guarantee or SLO, or as the
+user specifying their requirements.  Around evictions, there are typically three
+parameters:
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
+
+## Implementations
+
+In order of increasing complexity (and benefit).
+
+### Drain one node
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
+
+### One-shot drain of multiple nodes
+
+In this case, we try to meet some specific service goals specified by the admin
+on the command line.  The specification of these parameters only when we run the
+tool lets us provide the possibility of safety while avoiding questions of where
+to configure and store such specification (treated further down).
+
+```
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
```

Comment:
I was assuming that the "drain multiple" case meant "drain one at a time and then return to service, and then go on to the next one".  But, maybe you didn't mean that.  Draining machines in blocks is useful if you have a lot of machines or need to work on several at once (e.g. drain a rack).  So, I think "drain multiple in parallel" is a pretty useful function.  The "drain multiple in series" is less obviously useful since I don't see yet how you integrate that with the other steps of the maintenance.